### PR TITLE
todo-crud-completed

### PR DESCRIPTION
**Pull request by anees-rehman-brainx**

### CODE REVIEW

File: controller/todo-controller.js [hunk 1]
```diff
@@ -1,39 +1,41 @@
...
-            userId : userId,
-            todoId : Math.floor(Math.random() * 100) + userId,
-            message : message,
-            createdAt : date
+   try {
+        const {title, message} = req.body;
+        const {userId} = req.params;
+
+        if(!title || !message){
+            res.status(400).json("All feilds required");
```
Comment (ali-iqbal-brainx) [inline controller/todo-controller.js]: status should be 406 here

File: controller/todo-controller.js [hunk 2]
```diff
@@ -1,39 +1,41 @@
...
         }
-        
-    } catch (error) {
-        res.status(404).json(error)
-    }
+
+        res.status(201).json({message : "todo created successfully", payload : newTodo});
+
+   } catch (error) {
+        res.status(400).json(error);
```
Comment (ali-iqbal-brainx) [inline controller/todo-controller.js]: status code here should be 500

File: controller/todo-controller.js [hunk 3]
```diff
@@ -1,39 +1,41 @@
...
-        const date = new Date();
-
-        const todo ={
-            userId : userId,
-            todoId : Math.floor(Math.random() * 100) + userId,
-            message : message,
-            createdAt : date
+   try {
+        const {title, message} = req.body;
+        const {userId} = req.params;
```
Comment (ali-iqbal-brainx) [inline controller/todo-controller.js]: get userId from token instead params

File: controller/todo-controller.js [hunk 4]
```diff
@@ -1,39 +1,41 @@
...
 }
 
 // get all todos
 const getAllTodos = async (req, res) => {
-
     try {
-        const todos = todoService.getAllTodos();
+        const todos = await todoService.getAllTodos();
+
+        if(!todos){
```
Comment (ali-iqbal-brainx) [inline controller/todo-controller.js]: todo is an array use .length

File: controller/todo-controller.js [hunk 5]
```diff
@@ -45,7 +47,13 @@ const getAllTodos = async (req, res) => {
 const getTodoByUserId = async (req, res) => {
     try {
         const {userId} = req.params;
-        const todos = todoService.getTodosByUserId(userId);
+        const todos = await todoService.getTodosByUserId(userId);
+
+        if(!todos){
```
Comment (ali-iqbal-brainx) [inline controller/todo-controller.js]: todo is an array

File: controller/todo-controller.js [hunk 6]
```diff
@@ -69,17 +83,22 @@ const getTodoByTodoId =async (req, res) => {
...
-        const {userId, todoId} = req.params;
-        const {message} = req.body;
+        const {todoId} = req.params;
+        const {title, message} = req.body;
 
-        if(todoService.validateTodoId(todoId)){
-            todoService.updateTodo(todoId, message);
-            res.status(200).json("Todo updated Successfully");
+        if(!title || !message){
+            res.status(400).json("All feilds required");
```
Comment (ali-iqbal-brainx) [inline controller/todo-controller.js]: use proper status codes (406)

File: controller/user-controller.js [hunk 7]
```diff
@@ -3,192 +3,250 @@ const bcrypt = require('bcrypt');
...
-    const salt = await bcrypt.genSalt(10);
-    const hashedPassword = await bcrypt.hash(password, salt);
-    
-    const user = {
-        username : username,
-        email : email,
-        password : hashedPassword
-    }
+        if (password !== confirmPassword) {
+            res.status(403).json("Password not same");
```
Comment (ali-iqbal-brainx) [inline controller/user-controller.js]: use 400 status code here

File: controller/user-controller.js [hunk 8]
```diff
@@ -3,192 +3,250 @@ const bcrypt = require('bcrypt');
...
-            _id : savedUser._id,
-            username : savedUser.username,
-            email : savedUser.email,
-            
-        }
-        res.status(201).json({message : "User created Successfully", user : payload});
+        //Encrypting the password
+        const salt = await bcrypt.genSalt(10);
+        const hashedPassword = await bcrypt.hash(password, salt);
+        const savedUser = await userService.addUser({ username, email, password: hashedPassword });
```
Comment (ali-iqbal-brainx) [inline controller/user-controller.js:36]: this can be done only with a single query

File: controller/todo-controller.js (hunk 3 above)
Comment (anees-rehman-brainx) [reply controller/todo-controller.js]: In middelware userId is extracted from token and saved as req.params

File: controller/user-controller.js [hunk 9]
```diff
@@ -3,192 +3,250 @@ const bcrypt = require('bcrypt');
...
-                if(userToken){
-                    res.status(201).json({message : "Login Successfull", token : userToken.token})
-                }
 
-            } catch (error) {
-                res.json(error);
-            }
-        }
+        res.setHeader("access-control-expose-headers", "access-token")
+            .header("access-token", existedUser.token)
```
Comment (ali-iqbal-brainx) [inline controller/user-controller.js]: token should be generated at every login

File: controller/user-controller.js [hunk 10]
```diff
@@ -3,192 +3,250 @@ const bcrypt = require('bcrypt');
...
+        if (!oldPassword || !newPassword || !confirmNewPassword) {
+            res.status(406).json("All feilds are required...");
+            return;
+        }
 
-    if(!isUserExist){
-        res.status(404).json("Email does't exist..");
-        return;
-    }
+        if (newPassword != confirmNewPassword) {
```
Comment (ali-iqbal-brainx) [inline controller/user-controller.js]: use !== instead !=

File: controller/user-controller.js [hunk 11]
```diff
@@ -3,192 +3,250 @@ const bcrypt = require('bcrypt');
...
-        res.status(404).json("Incorrect old password..");
-        return;
-    }
+        const existedUser = await userService.getUserById(userId);
 
-    if(newPassword != confirmNewPassword){
-        res.status(404).json("New Password does't match");
-        return;
-    }
+        if (existedUser) {
```
Comment (ali-iqbal-brainx) [inline controller/user-controller.js]: invalid condition

File: controller/user-controller.js [hunk 12]
```diff
@@ -3,192 +3,250 @@ const bcrypt = require('bcrypt');
...
     // creating a unique token for password reset link
-    const secretKey = process.env.JWT_SECRET + user.password;
-    const payload = {email : user.email, id : user.id};
-    const token = jwt.sign(payload, secretKey, {expiresIn : '15m'});
+    const token = jwtService.generateAccessToken(user);
 
     // Password resett link that is to be mailed to user 
-    const link = `http://localhost:${process.env.port}/user/reset-password/${user.userId}/${token}`;
-
+    const link = `${process.env.HOST}:${process.env.port}/user/reset-password/${user._id}/${token}`;
```
Comment (ali-iqbal-brainx) [inline controller/user-controller.js]: port no should be in HOST

File: controller/user-controller.js [hunk 13]
```diff
@@ -3,192 +3,250 @@ const bcrypt = require('bcrypt');
...
+            user : 'aneesrehmanur@gmail.com',
+            pass : process.env.PASSWORD
+        }
+    });
+
+    let mailDetails = {
+        from : process.env.USERNAME,
+        to : user.email,
+        subject : "Password Resett Link",
+        text : `<a href = ${link}>click to Reset Password </a>`
```
Comment (ali-iqbal-brainx) [inline controller/user-controller.js]: message type should be html instead of text

File: controller/user-controller.js [hunk 14]
```diff
@@ -3,192 +3,250 @@ const bcrypt = require('bcrypt');
...
+            return res.status(500).json("Unexpected error occured");
+        }
 
-    
-    } catch (error) {
+        return res.status(201).json("Password updated successfully..");
+
+    }
+    catch (error) {
         res.json(error);
```
Comment (ali-iqbal-brainx) [inline controller/user-controller.js]: send proper status code for error

File: middelwares/authentication.js [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+                if (err) {
+                    return response
+                        .status(401)
+                        .json({ error: "Token is not valid!" });
+                }
+
+                const dbUser = await userService.getUserById(user._id);
+                console.log("User :", dbUser);
+
+                // console.log("db user expired token array :", dbUser.expiredTokens);
```
Comment (ali-iqbal-brainx) [inline middelwares/authentication.js]: remove commented code

File: middelwares/authentication.js [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+    console.log("In verify auth");
+    const authHeader = request.headers.access_token;
+    console.log(authHeader)
+    if (authHeader) {
+        const token = authHeader.split(".")[1];
+
+        console.log("token :", token);
+
+        jwtService.verifyToken(
+            authHeader,
```
Comment (ali-iqbal-brainx) [inline middelwares/authentication.js]: authHeader should be replaced with token

File: middelwares/authentication.js [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+                        .status(404)
+                        .json({ error: "User not found" });
+
+                if (dbUser.token === null) {
+                    return response
+                        .status(401)
+                        .json({ error: "Session token not found" });
+                }
+
+                if (dbUser.token !== authHeader) {
```
Comment (ali-iqbal-brainx) [inline middelwares/authentication.js]: invalid condition

File: modals/todo-modal.js [hunk 18]
```diff
@@ -1,9 +1,10 @@
 const mongoose = require('mongoose');
+const status = require('../constants/shared')
 
 const todoSchema = mongoose.Schema({
  
     userId : {
-        type : mongoose.SchemaTypes.ObjectId,
+        type : String,
```
Comment (ali-iqbal-brainx) [inline modals/todo-modal.js]: object id type must be an objectId not string

File: routes/user-route.js [hunk 19]
```diff
@@ -1,13 +1,13 @@
 const express = require('express');
 const userController = require('../controller/user-controller')
-
+const authMiddelware = require("../middelwares/authentication")
 const router = express.Router();
 
 router.post("/register", userController.register);
 router.post("/login", userController.login);
-router.get("/change-password", userController.changePassword);
+router.get("/change-password", authMiddelware.verifyAuth, userController.changePassword);
```
Comment (ali-iqbal-brainx) [inline routes/user-route.js]: router type should be put here

File: middelwares/authentication.js (hunk 16 above)
Comment (anees-rehman-brainx) [reply middelwares/authentication.js]: Not working with token

File: service/todo-service.js [hunk 20]
```diff
@@ -1,52 +1,52 @@
...
+    const todos = await todoModal.find();
     return todos;
 }
 
 // getTodosByUserId
-const getTodosByUserId = (userId) => {
-    const userTodos = todos.filter(todo => todo.userId === userId);
-    return userTodos;
+const getTodosByUserId = async (userId) => {
+    const todo = await todoModal.findOne({userId : userId});
```
Comment (ali-iqbal-brainx) [inline service/todo-service.js]: query should be find() here

File: service/user-service.js [hunk 21]
```diff
@@ -1,73 +1,46 @@
...
 
-
 // change password
-const updatePassword = (userId, newPassword) => {
-    const index = users.findIndex(user => user.userId === userId );
-    if(index != -1){
-        users[index].password = newPassword;
-    }
+const updatePassword = async (userId, newPassword) => {
+ const updatedUser = await userModal.findByIdAndUpdate(userId, user);
```
Comment (ali-iqbal-brainx) [inline service/user-service.js]: error here

**Review by ali-iqbal-brainx — CHANGES_REQUESTED** [review]
-get userId from token instead params
-use proper status codes for every error(5)
-todo is an array use .length(2)
-token should be generated at every login
-use !== instead !=
-invalid condition
-port no should be in HOST
-message type should be html instead of text
-remove commented code
-authHeader should be replaced with token
-object id type must be an objectId not string
-router type for update should be put
-use proper queries of mongoDB

**Pull request merged**